tb_plugin/torch_tb_profiler/profiler/op_tree.py

```diff
@@ -0,0 +1,329 @@
+
+    def __init__(self, tid2list, tid2zero_rt_list, corrid_to_device, fwd_bwd_map):
+        self.tid2list = tid2list
+        self.tid2zero_rt_list = tid2zero_rt_list
+        self.corrid_to_device = corrid_to_device
+        self.fwd_bwd_map = fwd_bwd_map
+        self.main_tid = None
+        self.tid2tree = None
+
+    def build_tree(self):
+        '''Construct the BackwardNode and replace the original backward nodes
+        '''
+        self.tid2tree = self._build_tree(self.tid2list, self.tid2zero_rt_list, self.corrid_to_device)
+        self._set_main_tid()
+
+        modules, backward_nodes = self._get_modules()
+        if not modules or not backward_nodes:
+            return self.tid2tree
+
+        _, ts2parent = OpTreeBuilder._get_node_parents(backward_nodes)
+        agg_nodes = OpTreeBuilder._group_backward_nodes(backward_nodes)
+        fwd_bwd_root = self._get_backward_roots(self.fwd_bwd_map, ts2parent, agg_nodes)
+        if len(agg_nodes) > 0:
+            logger.warning("some nodes cannot find forward nodes")
+
+        backward_modules = []
+        for module in modules:
+            OpTreeBuilder._build_backward_module(module, None, fwd_bwd_root, backward_modules)
+        OpTreeBuilder._insert_backward_modules(self.tid2tree[self.main_tid], backward_modules)
+        self.tid2tree = {tid: root for tid, root in self.tid2tree.items() if len(root.children) > 0}
+
+        return self.tid2tree
+
+    def _build_tree(self, tid2list, tid2zero_rt_list, corrid_to_device):
+        tid2tree = {}
+
+        staled_device_nodes = []
+        for _, device_nodes in corrid_to_device.items():
+            staled_device_nodes.extend([n for n in device_nodes if n.type == EventTypes.KERNEL])
+
+        for tid, op_list in tid2list.items():
+            zero_rt_list = tid2zero_rt_list[tid] if tid in tid2zero_rt_list else []
+            # Note that when 2 start_time are equal, the one with bigger end_time should be ahead of the other.
+            op_list.sort(key=lambda x: (x.start_time, -x.end_time))
+            main_tid = any([op.name.startswith("ProfilerStep#") for op in op_list])
+            if main_tid:
+                # only append the staled device nodes into main thread
+                self.main_tid = op_list[0].tid
+                root_node = self._build_tree_internal(op_list, zero_rt_list, tid, staled_device_nodes)
+            else:
+                root_node = self._build_tree_internal(op_list, zero_rt_list, tid, [])
+            tid2tree[int(tid)] = root_node
+
+        return tid2tree
+
+    def _set_main_tid(self):
+        if self.main_tid is None and self.tid2tree:
+            if len(self.tid2tree) == 1:
+                self.main_tid = next(iter(self.tid2tree))
+            else:
+                # there are multiple tids
+                backward_tid = self._find_backward_tid()
+                backward_tid = None
```

Comment:
why find_backward_tid and then set backward_tid to None?

Reply:
good catch, it is a bug